### PR TITLE
support(menu): Fix cycle up

### PR DESCRIPTION
**Pull request by gillchristian**
Solves #314 

**Problem**: When moving on the menu (`MenuPreviousItem` & `MenuNextItem`) going for last to first did work but going from first to last didn't.

**Solution**: Updated the helper used to get the new position on any of the actions that change menu position to cycle up and down.

---

There were no tests for the store so I added some to cover the cases of my changes. While writing the tests I found that `MenuPosition` could set the position out of range so also added the necessary changes to prevent that.

---

**NOTE**: this is the first time I work with Native Reason and also my first contribution to the project. I'm not sure the tests follow the approach the team likes so please let me know what could be improved :sweat_smile:

### CODE REVIEW

**Comment by gillchristian** [comment]
This is my first PR. I have a few questions general questions.

- Is there some format that should be followed for the branch names (since it's a fork maybe it's not relevant) and commit messages?
- Is there a format for the PR? Maybe is worth adding a template :smile: 

And also a question about this particular PR. I did not see any tests for `src/editor/Store`, should I tests for this changes? Is adding tests for `src/editor/Store` planned? If so maybe is worth adding and issue for that.

**Comment by Akin909** [comment]
@gillchristian thanks for the PR :+1: , re. commit messages and branching we don't have any strict rules around that.

Re the format of the PR there's some guidance re contributing in general like how to go about tackling a problem but if you mean the content of your pr message, nope no strict guidance.

I'm personally a fan on reducing friction as much as possible @bryphe might have some other thoughts.

Re. tests, I think we'd be all for any tests tests you add :rocket:

**Comment by gillchristian** [comment]
> if you mean the content of your pr message

Yes I meant the message.

> I think we'd be all for any tests tests you add

Great! I'm new to Reason and Revery, I'll check the other tests and see what I can add. Most probably I'll propose something here before adding anything.

**Comment by gillchristian** [comment]
This is ready for review :sweat_smile:

**Comment by lpalmes** [comment]
@gillchristian is a friend of mine @bryphe, so you can blindly merge the pr 😆

In all seriousness, i'm really happy that people are contributing here! (sorry for the spamy comment, i'm just happy about this)

**Comment by gillchristian** [comment]
I see tests failing on MacOS for a different suite. I don't have a Mac to reproduce. On my machine (Ubuhtu 18.04) all run fine so I cannot check if my changes did break something.

File: test/editor/Store/Helpers.re [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Model = Oni_Model;
+
+let onStateChange = (updater, initialState, action, onChange) => {
+  let (storeDispatch, _) =
+    Isolinear.Store.create(~initialState, ~updater, ());
+
+  action |> storeDispatch |> fst |> onChange;
```
Comment (bryphe) [inline test/editor/Store/Helpers.re]: Awesome! I like the way you abstracted this out to make it easy to reuse across tests.

File: test/editor/Store/MenuStoreConnectorTest.re [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+let commands = [cmd, cmd, cmd, cmd];
+
+let onStateChange = MenuStore.start() |> fst |> Helpers.onStateChange;
+
+let store =
+  describe("MenuStoreConnector", ({describe, _}) => {
+    describe("MenuPreviousItem", ({test, _}) => {
+      test("goes from 1st to last", ({expect}) =>
+        onStateChange(
```
Comment (bryphe) [inline test/editor/Store/MenuStoreConnectorTest.re]: Impressive work on these tests - thank you for getting `Store` tests started for us 😍

**Comment by bryphe** [comment]
Sorry for the delay on my end @gillchristian ! The change looks great - thank you for going the extra mile and figuring out how to wire up tests for the `Store` - that will be very useful as we focus on stability.

> In all seriousness, i'm really happy that people are contributing here! (sorry for the spamy comment, i'm just happy about this)

@lpalmes me too! 👍 Awesome that you guys know each other 😄 

> I see tests failing on MacOS for a different suite. I don't have a Mac to reproduce. On my machine (Ubuhtu 18.04) all run fine so I cannot check if my changes did break something.

I kicked off the build again to see if it's an intermittent failure. I'll check out on my Mac if it's still failing - otherwise, will bring this in once its green. Thank you very much for the contribution!

**Comment by gillchristian** [comment]
No worries. Whenever this can move forward is fine to me.

I'm happy you liked the tests, if there's anything that could be improved let me know. Now that I have an idea of how to write these tests I think I will work on increasing the coverage of `Store` tests.

**Comment by bryphe** [comment]
> No worries. Whenever this can move forward is fine to me.

Cool! I'm a bit short on time at the moment (heading to Paris for ReactEurope), but will look again when I come back. Unfortunately, this reproduced consistently on the build machine - I think there is some interaction with the textmate tests (which host their own service). 

Once we have a native version of `vscode-textmate` - https://github.com/onivim/oni2/issues/161 - these tests should, as a side effect, get more reliable too!

**Comment by gillchristian** [comment]
@bryphe it's been a while since I opened this PR, even the first release happened :tada: 

I rebased with master and there are two problems.

First I force pushed to my branch without your changes :face_with_head_bandage: 

Also I tried to run the tests again but I get an error when building, maybe it's better to ask in Discord about this :thinking: 

```
$ esy '@test' build
info building reason-libvim@github:onivim/reason-libvim#5c81c76@d41d8cd9
error: build failed with exit code: 1
  build log:
    # esy-build-package: building: reason-libvim@github:onivim/reason-libvim#5c81c76
    # esy-build-package: pwd: /home/gillchristian/.esy/3____________________________________________________________/b/reason_libvim-445ec134
    # esy-build-package: running: 'refmterr' 'dune' 'build' '-p' 'libvim'
        discover src/c_flags.sexp,src/cxx_flags.sexp,src/flags.sexp
    -L/home/gillchristian/.esy/3/i/libvim-8.10869.19-49bdd02b/lib
        ocamlopt src/vim.cmxs (exit 2)
    (cd _build/default && /home/gillchristian/.esy/3/i/ocaml-4.7.1004-a6ae015e/bin/ocamlopt.opt -w -40 -g -shared -linkall -I src -o src/vim.cmxs src/vim.cmxa)
    /usr/bin/ld: cannot find -lacl
    collect2: error: ld returned 1 exit status
    File "caml_startup", line 1:
    Error: Error during linking
    error: command failed: 'refmterr' 'dune' 'build' '-p' 'libvim' (exited with 1)
    esy-build-package: exiting with errors above...

  building reason-libvim@github:onivim/reason-libvim#5c81c76
esy: exiting due to errors above
```

**Comment by CrossR** [comment]
If you are on Ubuntu, to build now you also need the `libacl1-dev` package, which should fix that linker error.

**Comment by bryphe** [comment]
Unfortunately this is pretty out-of-date and seems like it would be superseded by the work @glennsl is doing on the menu - will close out.

**Pull request closed without merging**